File: backend/app/api/v1/endpoints/email_templates.py
"""
Email template endpoints
"""
from fastapi import APIRouter, Depends, HTTPException, status, Query
from typing import List, Optional
from app.middleware.auth import get_current_admin
from app.db.database import get_database
from bson import ObjectId
from datetime import datetime

router = APIRouter()


@router.get("/")
async def get_email_templates(
    organization: Optional[str] = Query(None),
    type: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_admin)
):
    """Get all email templates (admin only)"""
    db = await get_database()
    query = {}
    
    if organization:
        query["organization"] = ObjectId(organization)
    if type:
        query["type"] = type
    
    cursor = db.email_templates.find(query)
    templates = await cursor.to_list(length=100)
    
    result = []
    for template in templates:
        template["id"] = str(template["_id"])
        if template.get("organization"):
            template["organization"] = str(template["organization"])
        del template["_id"]
        result.append(template)
    
    return result


@router.get("/{template_id}")
async def get_email_template(template_id: str, current_user: dict = Depends(get_current_admin)):
    """Get email template by ID (admin only)"""
    db = await get_database()
    
    try:
        template = await db.email_templates.find_one({"_id": ObjectId(template_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Template not found"
        )
    
    if not template:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Template not found"
        )
    
    template["id"] = str(template["_id"])
    if template.get("organization"):
        template["organization"] = str(template["organization"])
    del template["_id"]
    
    return template


@router.post("/")
async def create_email_template(template_data: dict, current_user: dict = Depends(get_current_admin)):
    """Create email template (admin only)"""
    db = await get_database()
    
    if template_data.get("organization"):
        template_data["organization"] = ObjectId(template_data["organization"])
    
    template_data["created_at"] = datetime.utcnow()
    
    result = await db.email_templates.insert_one(template_data)
    template_data["id"] = str(result.inserted_id)
    if template_data.get("organization"):
        template_data["organization"] = str(template_data["organization"])
    del template_data["_id"]
    
    return template_data


@router.put("/{template_id}")
async def update_email_template(
    template_id: str,
    template_data: dict,
    current_user: dict = Depends(get_current_admin)
):
    """Update email template (admin only)"""
    db = await get_database()
    
    if template_data.get("organization"):
        template_data["organization"] = ObjectId(template_data["organization"])
    
    await db.email_templates.update_one(
        {"_id": ObjectId(template_id)},
        {"$set": template_data}
    )
    
    template = await db.email_templates.find_one({"_id": ObjectId(template_id)})
    template["id"] = str(template["_id"])
    if template.get("organization"):
        template["organization"] = str(template["organization"])
    del template["_id"]
    
    return template


@router.delete("/{template_id}")
async def delete_email_template(template_id: str, current_user: dict = Depends(get_current_admin)):
    """Delete email template (admin only)"""
    db = await get_database()
    
    await db.email_templates.delete_one({"_id": ObjectId(template_id)})
    
    return {"message": "Email template deleted successfully"}


@router.post("/{template_id}/preview")
async def preview_email_template(template_id: str, current_user: dict = Depends(get_current_admin)):
    """Preview email template (admin only)"""
    db = await get_database()
    
    template = await db.email_templates.find_one({"_id": ObjectId(template_id)})
    
    if not template:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Template not found"
        )
    
    # TODO: Render template with sample data
    return {
        "html": template.get("htmlBody", ""),
        "subject": template.get("subject", "")
    }

File: server/routes/emailTemplates.js
/**
 * Email Template Management Routes
 * Admin Only
 */

import express from 'express'
import EmailTemplate from '../models/EmailTemplate.js'
import { protect, admin } from '../middleware/auth.js'

const router = express.Router()

// All routes require admin access
router.use(protect, admin)

/**
 * @route   GET /api/email-templates
 * @desc    Get all email templates
 * @access  Private/Admin
 */
router.get('/', async (req, res) => {
  try {
    const { organization, type } = req.query
    const query = {}

    if (organization) {
      query.organization = organization
    } else {
      // Show global templates if no org specified
      query.$or = [
        { organization: null },
        { organization: { $exists: false } },
      ]
    }

    if (type) {
      query.type = type
    }

    const templates = await EmailTemplate.find(query)
      .populate('organization', 'name')
      .populate('createdBy', 'name email')
      .sort({ createdAt: -1 })

    res.json(templates)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   GET /api/email-templates/:id
 * @desc    Get email template by ID
 * @access  Private/Admin
 */
router.get('/:id', async (req, res) => {
  try {
    const template = await EmailTemplate.findById(req.params.id)
      .populate('organization', 'name')
      .populate('createdBy', 'name email')

    if (!template) {
      return res.status(404).json({ message: 'Template not found' })
    }

    res.json(template)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   POST /api/email-templates
 * @desc    Create email template
 * @access  Private/Admin
 */
router.post('/', async (req, res) => {
  try {
    const { name, type, subject, htmlBody, textBody, organization, variables } = req.body

    if (!name || !type || !subject || !htmlBody) {
      return res.status(400).json({ message: 'Name, type, subject, and htmlBody are required' })
    }

    const templateData = {
      name,
      type,
      subject,
      htmlBody,
      textBody: textBody || '',
      organization: organization || null,
      variables: variables || [],
      createdBy: req.user._id,
    }

    const template = await EmailTemplate.create(templateData)

    res.status(201).json(template)
  } catch (error) {
    if (error.code === 11000) {
      return res.status(400).json({ message: 'Template with this type and organization already exists' })
    }
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   PUT /api/email-templates/:id
 * @desc    Update email template
 * @access  Private/Admin
 */
router.put('/:id', async (req, res) => {
  try {
    const { name, subject, htmlBody, textBody, isActive, variables } = req.body

    const template = await EmailTemplate.findById(req.params.id)

    if (!template) {
      return res.status(404).json({ message: 'Template not found' })
    }

    if (name) template.name = name
    if (subject) template.subject = subject
    if (htmlBody) template.htmlBody = htmlBody
    if (textBody !== undefined) template.textBody = textBody
    if (isActive !== undefined) template.isActive = isActive
    if (variables) template.variables = variables

    await template.save()

    res.json(template)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   DELETE /api/email-templates/:id
 * @desc    Delete email template
 * @access  Private/Admin
 */
router.delete('/:id', async (req, res) => {
  try {
    const template = await EmailTemplate.findById(req.params.id)

    if (!template) {
      return res.status(404).json({ message: 'Template not found' })
    }

    await template.deleteOne()

    res.json({ message: 'Template deleted successfully' })
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   POST /api/email-templates/:id/preview
 * @desc    Preview email template with sample data
 * @access  Private/Admin
 */
router.post('/:id/preview', async (req, res) => {
  try {
    const template = await EmailTemplate.findById(req.params.id)

    if (!template) {
      return res.status(404).json({ message: 'Template not found' })
    }

    // Generate sample data based on template type
    const sampleData = generateSampleData(template.type)

    // Render template (simplified - in production, use proper templating engine)
    let rendered = template.htmlBody
    Object.keys(sampleData).forEach(key => {
      const regex = new RegExp(`\\{\\{${key}\\}\\}`, 'g')
      rendered = rendered.replace(regex, JSON.stringify(sampleData[key]))
    })

    res.json({ html: rendered, subject: template.subject })
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * Generate sample data for template preview
 */
const generateSampleData = (type) => {
  const baseData = {
    date: new Date().toLocaleDateString(),
    period: 'Sample',
  }

  switch (type) {
    case 'daily-open-tickets':
      return {
        ...baseData,
        tickets: [
          { ticketId: 1001, title: 'Sample Ticket 1', priority: 'high', department: { name: 'IT Support' }, assignee: { name: 'John Doe' }, dueDate: new Date(), createdAt: new Date() },
          { ticketId: 1002, title: 'Sample Ticket 2', priority: 'medium', department: { name: 'HR' }, assignee: null, dueDate: null, createdAt: new Date() },
        ],
      }
    case 'daily-report':
      return {
        ...baseData,
        totalCreated: 10,
        totalOpen: 5,
        totalResolved: 3,
        slaBreached: 1,
        departmentSummary: [{ departmentName: 'IT Support', count: 5 }, { departmentName: 'HR', count: 3 }],
      }
    case 'weekly-report':
      return {
        ...baseData,
        startDate: 'Jan 01, 2024',
        endDate: 'Jan 07, 2024',
        totalCreated: 50,
        resolved: 30,
        unresolved: 20,
        slaCompliant: 45,
        slaBreached: 5,
        technicianPerformance: [{ name: 'John Doe', total: 10, resolved: 8 }],
        topIssues: [{ _id: 'Hardware Issue', count: 10 }],
      }
    case 'monthly-report':
      return {
        ...baseData,
        month: 'January 2024',
        totalCreated: 200,
        departmentTrends: [{ departmentName: 'IT Support', count: 100 }],
        slaViolations: 10,
        slaComplianceRate: 95,
        technicianProductivity: [{ name: 'John Doe', total: 50, resolved: 45, resolutionRate: 90 }],
        recurringIssues: [{ _id: 'Hardware Issue', count: 20 }],
      }
    default:
      return baseData
  }
}

export default router

